docs/DMS/PRIMARY-DATA-STORAGE.md

```diff
@@ -25,3 +25,268 @@ Braindump:
+
+(**Maybe provide both options?**)
+
+The query table schema can be pre-generated of course, but also the JSON Paths to the queryable elements, and even the SQL insert statements -- though they'd have to target all supported databases.
+
+#### Security
+
+```mermaid
+erDiagram
+    Documents {
+        Guid referential_id PK "Hash of identity"
+        Guid document_uuid "API resource id, indexed"
+        string project_name "Example: Ed-Fi (for DS)"
+        string resource_name "Example: Student"
+        string resource_version "Example: 5.0.0"
+        JSON edfi_doc "The document"
+    }
+    StudentSchoolAssociationSecurity ||--|| Documents : ""
+    StudentSchoolAssociationSecurity {
+        bigint id PK "Auto-increment, otherwise unused"
+        Guid referential_id FK "SSA document being indexed"
+        string student_usi "Student unique id in this SSA document"
+        string school_id "School id in this SSAdocument"
+        string et_cetera
+    }
+```
+
+We expect that security can be handled the same way as queries, with sidecar tables generated per resource with the fields relevant to security extracted into columns.
+
+#### Issues
+
+Let's consider some downsides to this approach and how we can mitigate them. The most obvious is that all resources are stored in a single table which presents a inconvenience when you only want to see specific resources. A solution to that would be to create a view per entity, where the views just select out the JSON for entities of a particular resource type. For example, there would be a view `edfi.schools` which just does a `SELECT edfi_doc from Documents where resource_name = "School"`.
+
+Another potential downside is partitioning itself. However, designing for partitioning up front is better than adding on later, and there are resources for attendance and assessments with potential to need partitioning on their own. Hash partitioning is conceptually straightforward and operationally you just decide on the number of partitions to hash into.
+
+(**Hash-based partitioning is easy to declare in PostgreSQL, need to research more on SQL Server side**)
+
+Probably the biggest potential downside is a common SQL server community concern around using Guids as a primary key.
+
+(**There is a strong community bias towards sequential numbers as primary keys. It will be important to demonstrate that using a randomized uuid as a primary key can work just fine, but definitely requires different index maintenance from a sequential primary key. We will need to provide technical guidance on how index maintenance differs between tables with sequential IDs and tables with randomized IDs. We will need to provide real world testing showing that randomized IDs work very well so long as index maintenance is done properly.** [This presentation](https://www.youtube.com/watch?v=nc4CMo7VSPo) **is an excellent starting point.**)
```

Comment:
This is a good article. It actually could be tweaked into "How to use GUIDs..." instead of "Careful with GUIDs...". It describes how the real issue is that random GUID keys need to be managed differently than sequential keys. He's just a bit reluctant to deviate from the sequential key pattern.

In the section on "Random GUID keys", he's exactly right that a full data page is a terrible place for a new key to be inserted. With the sequential key pattern, you almost never run into that because you are only appending to the last page with a goal fill-factor of 100%. (The downside is that rapid index inserts create hot page contention.) Anyway, here your index maintenance goal is to pack everything to 100%.

However, if your plan is to randomly distribute keys across data pages, you would be **insane** to always pack the pages full! As he notes, full pages mean a page split on every insert.

He's actually got the answer to the problem in his "Fillfactor to the rescue?" section. The answer is easy, just run at a lower fill-factor. Now the inserts are randomly distributed across pages with no splitting, and with the added benefit of faster inserts due to no hot page contention. The index maintenance goal is now to maintain a fill-factor that allows room to avoid splitting between maintenance sessions. (Interestingly, because the amount of total free space grows as the number of pages grow, over time you could actually raise the fill-factor if you wanted to.)

Yes, this scheme requires more disk/memory space than sequential keys but it is also better suited to handle our reference validation and partitioning needs.